### PR TITLE
feat: create Auro button component

### PR DESCRIPTION
**Pull request by gusnaughton**
# Alaska Airlines Pull Request

Introduces the `<auro-button>` component with `<auro-button-light>` for Light DOM support. Auro-button introduces `ondark` functionality to the Auro design system.

Original Orion documentation has been preserved in the `./docs/ODS_README.md` file.

## Summary:

This pull request adds support for the upcoming Auro design system while maintaining prior Orion button support.

## Type of change:

- [x] New capability 

## Checklist:

- [x] My update follows the CONTRIBUTING guidelines of this project
- [x] I have performed a self-review of my own update

**By submitting this Pull Request, I confirm that my contribution is made under the terms of the Apache 2.0 license.**

_Pull Requests will be evaluated by their quality of update and whether it is consistent with the goals and values of this project. Any submission is to be considered a conversation between the submitter and the maintainers of this project and may require changes to your submission._ 

**Thank you for your submission!**<br>
-- Orion Design System Team

### CODE REVIEW

**Comment by gusnaughton** [comment]
Review items:

- Discussion on whether we should have two separate componentbase files (one for orion, one for auro) or if we want to split the common functionality across both of them (such as defining attributes). I'm keen to keep them separate, but I'd like to actually evaluate the potential risk of breaking changes for the orion-button before making this decision.

- Should we use the sassPartials pattern again for storing common styles between auro-button and auro-button-ld? In the current branch, the styles are largely duplicated which does not feel clean to me.

- On L77 of auro-button.js, I implemented a basic input validation check on which type of button to render. Do we want to throw a console error if an invalid value is passed here?

- In this PR, we modified the sassPartials folder to use the non-deprecated auro_transition. Should we refactor out all references to deprecated tokens in Orion in this PR? Should we revert this change to keep the scope _just_ Auro?

- Do we currently have a clean way of displaying on-dark components in the demo? I haven't spent much time in figuring out a clean way to display this. If we haven't done this before, I will investigate.

- [ ] Need to write documentation on Auro

File: src/auro-component-base.js [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+    this.addEventListener('touchstart', function() {
+      this.classList.add('is-touching');
+    });
+  }
+
+  static get properties() {
+    return {
+      autofocus:        { type: Boolean },
+      formnovalidate:   { type: Boolean },
+      isactive:         { type: Boolean },
```
Comment (blackfalcon) [inline src/auro-component-base.js]: not supported; remove

File: src/auro-component-base.js [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+      ondark:           { type: Boolean },
+      responsive:       { type: Boolean },
```
Comment (blackfalcon) [inline src/auro-component-base.js]: not supported; remove

File: src/auro-component-base.js [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+      reverse:          { type: Boolean },
```
Comment (blackfalcon) [inline src/auro-component-base.js]: not supported; remove

File: src/style-auro-ld.scss [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+    &:hover {
+      @include auro_transition(all, 150ms, ease-in-out);
+      
+      background-color: var(--auro-color-ui-hover-on-light);
+      border: 1px solid var(--auro-color-ui-hover-on-light);
+    }
+
+    &--secondary {
+      &:hover {
+        @include auro_transition();
```
Comment (blackfalcon) [inline src/style-auro-ld.scss]: unnecessary; hover will be inherited from auro-button. When defining modifiers, you only need to define the modification from the base selector.

File: src/style-auro-ld.scss [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+
+        background-color: var(--auro-color-background-lighter);
+        border: 1px solid var(--auro-color-ui-default-on-light);
+      }
+    }
+
+    &--tertiary {
+      &:hover {
+        @include auro_transition();
```
Comment (blackfalcon) [inline src/style-auro-ld.scss]: unnecessary; hover will be inherited from auro-button. When defining modifiers, you only need to define the modification from the base selector.

File: src/style-auro-ld.scss [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+
+        background-color: var(--auro-color-background-lighter);
+        border-color: transparent;
+      }
+    }
+
+    &Ondark {
+      &:hover {
+        @include auro_transition(all, 150ms, ease-in-out);
```
Comment (blackfalcon) [inline src/style-auro-ld.scss]: unnecessary; hover will be inherited from auro-button. When defining modifiers, you only need to define the modification from the base selector.

File: src/style-auro-ld.scss [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+        &:hover {
+          background-color: rgba(255, 255, 255, 0.15);
+        }
+      }
+    }
+  }
+}
+
+// light DOM support for focus-visible
+auro-button-light .focus-visible.auro-button {
```
Comment (blackfalcon) [inline src/style-auro-ld.scss]: 2 things. 

Is this right? In the light DOM, isn't the focus-visible selector on the auro-button-light element? And then .auro-button would be a descendent selector?

File: src/style-auro-ld.scss [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+  @import './style-auro.scss';
+}
+
+// light DOM support for hover states
+auro-button-light:not(.is-touching) {
+  .auro-button {
+    &:hover {
+      @include auro_transition(all, 150ms, ease-in-out);
+      
+      background-color: var(--auro-color-ui-hover-on-light);
```
Comment (blackfalcon) [inline src/style-auro-ld.scss]: It's bugging me that these are duplicated from the main stylesheet. This is why I had to make all those hover states mixins with ods-button. 

I really want to find a better way to manage this interaction between shadow and light DOM elements.

File: src/style-auro-ld.scss [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+// light DOM support for focus-visible
+auro-button-light .focus-visible.auro-button {
+    box-shadow: inset 0 0 0 1px var(--auro-color-border-focus-on-dark), inset 0 0 0 3px var(--auro-color-background-lightest);
+
+    &Ondark {
+      background-color: var(--auro-color-ui-default-on-dark);
+      box-shadow: inset 0 0 0 1px var(--auro-color-ui-active-on-dark), inset 0 0 0 3px transparent;
+      
+      &--secondary {
+        background-color: rgba(255, 255, 255, 0.15);
```
Comment (blackfalcon) [inline src/style-auro-ld.scss]: In the other sass file you made this a variable?

File: src/style-auro.scss [hunk 10]
```diff
@@ -0,0 +1,178 @@
...
+// general WCSS imports
+@import "./node_modules/@alaskaairux/orion-web-core-style-sheets/dist/breakpoints";
+@import "./node_modules/@alaskaairux/orion-web-core-style-sheets/dist/core";
+@import "./node_modules/@alaskaairux/orion-web-core-style-sheets/dist/animation";
+
+@import "./node_modules/@alaskaairux/orion-design-tokens/dist/tokens/SCSSVariableMap";
+
+$auro-inset-spacing-options: lg;
+$auro-inset-directions:'--squish';
+$color-base-black-opacity-15: rgba(255, 255, 255, 0.15);
```
Comment (blackfalcon) [inline src/style-auro.scss]: define this AFTER the imports. 

Lines 19-20 are to influence the output of the import on line 22. 

Line 21 has nothing to do with that.

File: src/style-auro.scss [hunk 11]
```diff
@@ -0,0 +1,178 @@
...
+
+  &Ondark {
+    background-color: var(--auro-color-ui-default-on-dark);
+    border: none;
+    color: var(--auro-color-text-primary-on-dark);
+    
+    // auro-buttonOndark is-touching
+    :host(:not(.is-touching)) & {
+      &:hover {
+        @include auro_transition(all, 150ms, ease-in-out);
```
Comment (blackfalcon) [inline src/style-auro.scss]: No need to repeat this property

Comment (blackfalcon) [inline src/style-auro.scss]: This property should not be required as it's inherited from the base selector.

File: src/style-auro.scss [hunk 12]
```diff
@@ -0,0 +1,178 @@
...
+
+  // auro-button--secondary
+  &--secondary {
+    background-color: var(--auro-color-background-lightest);
+    border: 1px solid var(--auro-color-ui-default-on-light);
+    color: var(--auro-color-text-link-on-light);
+
+    :host(:not(.is-touching)) & {
+      &:hover {
+        @include auro_transition();
```
Comment (blackfalcon) [inline src/style-auro.scss]: Should not be needed.

File: src/style-auro.scss [hunk 13]
```diff
@@ -0,0 +1,178 @@
...
+  
+  // auro-button--tertiary
+  &--tertiary {
+    background-color: transparent;
+    border-color: transparent;
+    color: var(--auro-color-text-link-on-light);
+
+    :host(:not(.is-touching)) & {
+      &:hover {
+        @include auro_transition();
```
Comment (blackfalcon) [inline src/style-auro.scss]: Should not be needed.

File: src/auro-component-base.js [hunk 14]
```diff
@@ -0,0 +1,96 @@
...
+    return html`<slot></slot>`;
+  }
+
+  render() {
+
+    const classes = {
+      'util_insetLg--squish': true,
+      'auro-button': true,
+      'auro-buttonOndark': this.ondark,
+      'auro-button--secondary': this.secondary && !this.ondark,
```
Comment (blackfalcon) [inline src/auro-component-base.js]: Thought we decided that the negative reference was overkill?

File: src/sassPartials/_shared_style.scss [hunk 15]
```diff
@@ -13,7 +13,7 @@
     $cta: !important;
   }
 
-  @include transition;
+  @include auro_transition;
```
Comment (blackfalcon) [inline src/sassPartials/_shared_style.scss]: This is being used by ods-button? If so, this should stay with the deprecated one.

File: src/style-auro-ld.scss (hunk 7 above)
Comment (gusnaughton) [reply src/style-auro-ld.scss]: This selector resolves to the following:
`
<auro-button-light>
  …
    <element class="focus-visible auro-button">
`
Example of placement in the DOM:
`
<auro-button-light tertiary content="Tertiary default state">     
      
      <button class="util_insetLg--squish auro-button auro-button--tertiary focus-visible">
        Tertiary default state
      </button>
</auro-button-light>
`

Should we evaluate moving focus-visible up to auro-button-light?

File: src/style-auro-ld.scss (hunk 8 above)
Comment (gusnaughton) [reply src/style-auro-ld.scss]: We briefly discussed this problem a couple days ago, but I'm having trouble remembering our conclusion. Should we move back towards the hover states mixin pattern? What would you propose aside from the mixins to solve this problem?

Comment (blackfalcon) [reply src/style-auro-ld.scss]: I am going to resolve this conversation as I don't have an answer right now. But this is something that needs to be refactored at a future date.

File: src/style-auro.scss [hunk 16]
```diff
@@ -0,0 +1,175 @@
...
+$color-base-black-opacity-15: rgba(255, 255, 255, 0.15);
+
+// handle focus-visible state
+:host(.focus-visible) {
+  .auro-button {
+    box-shadow: inset 0 0 0 1px var(--auro-color-border-focus-on-dark), inset 0 0 0 3px var(--auro-color-background-lightest);
+
+    // auro-buttonOndark
+    &Ondark {
+      background-color: var(--auro-color-ui-default-on-dark);
```
Comment (blackfalcon) [inline src/style-auro.scss]: Double-check your values here. The style is being applied, but the UI is not correct.

File: src/style-auro-ld.scss (hunk 7 above)
Comment (blackfalcon) [reply src/style-auro-ld.scss]: Oops, that's my bad. I forgot that the focus-visible appears on the button element in the light DOM version.

File: demo/index.html [hunk 17]
```diff
@@ -25,6 +27,52 @@
...
+          <h3>Shadow DOM element examples</h3>
+          <demo-snippet>
+            <template>
+              <auro-button>Default state</auro-button>
+              <auro-button secondary>Secondary Default state</auro-button>
+              <auro-button tertiary>Tertiary default state</auro-button>
+            </template>
+          </demo-snippet>
+    
+          <demo-snippet class="dark">
```
Comment (blackfalcon) [inline demo/index.html]: This selector is not doing anything? 

Also, should be named `onDark` to be consistent.

File: src/style-auro.scss (hunk 16 above)
Comment (gusnaughton) [reply src/style-auro.scss]: What specifically is broken on the UI?

File: demo/index.html (hunk 17 above)
Comment (gusnaughton) [reply demo/index.html]: Removed the dark class. As discussed, @blackfalcon will be looking at this as we cannot style demo-snippet in its current state.

File: src/style-auro.scss [hunk 18]
```diff
@@ -164,9 +172,11 @@ $color-base-black-opacity-15: rgba(255, 255, 255, 0.15);
       }
     }
 
-    &:active, &:hover {
-      background-color: var(--auro-color-ui-hover-on-dark);
-      border: 1px solid var(--auro-color-ui-hover-on-dark);
+    @media (hover: hover) {
+      &:active, &:hover {
```
Comment (blackfalcon) [inline src/style-auro.scss]: nit-pick: returns between selectors

```
&:active, 
&: hover {
```

File: src/style-auro.scss [hunk 19]
```diff
@@ -140,9 +146,11 @@ $color-base-black-opacity-15: rgba(255, 255, 255, 0.15);
       }
     }
 
-    &:active, &:hover {
-      background-color: var(--auro-color-background-lighter);
-      border-color: transparent;
+    @media (hover: hover) {
+      &:active, &:hover {
```
Comment (blackfalcon) [inline src/style-auro.scss]: please use hard returns

File: src/style-auro.scss [hunk 20]
```diff
@@ -180,44 +190,48 @@ $color-base-black-opacity-15: rgba(255, 255, 255, 0.15);
       border: 1px solid var(--auro-color-ui-default-on-dark);
       color: var(--auro-color-ui-default-on-dark);
   
-      &:active, &:hover {
-        background-color: $color-base-black-opacity-15;
+      @media (hover: hover) {
+        &:active, &:hover {
```
Comment (blackfalcon) [inline src/style-auro.scss]: ditto

File: src/style-auro.scss [hunk 21]
```diff
@@ -180,44 +190,48 @@ $color-base-black-opacity-15: rgba(255, 255, 255, 0.15);
       border: 1px solid var(--auro-color-ui-default-on-dark);
       color: var(--auro-color-ui-default-on-dark);
   
-      &:active, &:hover {
-        background-color: $color-base-black-opacity-15;
+      @media (hover: hover) {
```
Comment (blackfalcon) [inline src/style-auro.scss]: what is this?

**Comment by corobrie** [comment]
Design Approved - Sir Corey Lee O’Brien of the North

File: src/style-auro.scss (hunk 21 above)
Comment (gusnaughton) [reply src/style-auro.scss]: https://drafts.csswg.org/mediaqueries/#hover determines whether the browser supports the hover state

File: src/style-auro.scss [hunk 22]
```diff
@@ -0,0 +1,246 @@
...
+    width: auto;
+  }
+
+  // auro-button--secondary
+  &--secondary {
+    background-color: var(--auro-color-background-lightest);
+    border: 1px solid var(--auro-color-ui-default-on-light);
+    color: var(--auro-color-text-link-on-light);
+
+    :host(:not(.is-touching)) & {
```
Comment (blackfalcon) [inline src/style-auro.scss]: Given the new @media query, is this selector still needed?

Comment (gusnaughton) [reply src/style-auro.scss]: Good call. Refactored these references out.

**Pull request merged**